viite-backend/base/src/main/scala/fi/vaylavirasto/viite/model/AddrMRange.scala

```diff
@@ -0,0 +1,50 @@
+package fi.vaylavirasto.viite.model
+
+import fi.vaylavirasto.viite.util.ViiteException
+
+/**
+ * AddrMRange resembles start address to end address value range (meters) for a section of a road, in context of a [[RoadPart]].
+ * AddrMRange class provides basic validation for the numbers' correctness, and basic boolean functions to easily check the state,
+ * and compare AddrMRanges.
+ *
+ * If an AddrMRange has both its values 0, that is, AddrMRange(0,0), the AddrMRange is said to be undefined, and/or invalid.
+ *
+ * @throws ViiteException at construction, if the start, or end are out of acceptable values.
+ */
+case class AddrMRange (start: Long, end: Long)// extends Ordered[AddrMRange]
+{
+
+  val minAddrMZero = 0        // A metric address value for a road part cannot ever be negative.
+  val maxAddrM     = 100*1000 // Semi-random upper limit. One road part will never be over 100km ever.
+
+  // throws ViiteException if requirements not met
+  try {
+    require((start>=minAddrMZero && start <= maxAddrM), s"A start address must be between $minAddrMZero-$maxAddrM.")
+    require((  end>=minAddrMZero &&   end <= maxAddrM), s"An  end address must be between $minAddrMZero-$maxAddrM.")
+    // Calculations so far require leaving this - I think very substantial - restriction out.
+    //require((  start < end || (start==0 &&  end==0)), s"A start address (now $start) must be smaller than end address (now $end), or range must be undefined (0-0).")
```

Comment:
Joo uusilla linkeillä taitaa olla esim originalAddrMValuet aina 0 ja ennen laskentaa taitaa olla normaalit osoitearvot myös

Reply:
Jep, periaatteessa on. 👍 

Vaan osa laskennoista ja testeistä feilas kun tuota tein.
Käytin ainakin viikon sen ihmettelyyn, miten korjailisin testejä tai koodia niin että saisin ton mukaan, mutta ne päätyi tuonne liu'utus- ym. kaninkoloihin.

Kun saadaan liu'utuskoodi sisään, tuon lisäystä vaatimuksiin vois koettaa uudestaan; feilit tuli jostain niiltä hoodeilta. Liu'utuksen sisään laiton jälkeen saattais mennä läpikin jos hyvin käy.

Reply:
liu'utus koodi on jo mergetty 👍

Reply:
Oh, no ootas, ajelenpa uudestaan toi enabloituna!

Reply:
Edelleen failaa 7. Mut ei enää 12.
Mut ei siis voi vieläkän enabloida noita.

ProjectSectionCalculatorSpec:
- Test assignAddrMValues When adding New link sections to the existing ones Then Direction for those should always be the same as the existing section (which are for e.g. Transfer, Unchanged, Numbering) making the addressMValues be sequential *** FAILED ***
- Test assignAddrMValues When adding New link sections with minor discontinuity to the existing ones Then Direction for those should always be the same as the existing section (which are for e.g. Transfer, Unchanged, Numbering) making the addressMValues be sequential *** FAILED ***
DefaultSectionCalculatorStrategySpec:
- Test defaultSectionCalculatorStrategy.assignAddrMValues() When two track road has new links and terminations in the middle of the road Then addressess and geometries should be continuous. *** FAILED ***
ProjectServiceSpec:
- Test updateRoadwaysAndLinearLocationsWithProjectLinks When VIITE-2236 situation Then don't violate unique constraint on roadway_point table. *** FAILED ***
- Test defaultSectionCalculatorStrategy.updateRoadwaysAndLinearLocationsWithProjectLinks() When two track road with a new track 2 and partially new track 1 having new track at the end Then formed roadways should be continuous. *** FAILED ***
- Test defaultSectionCalculatorStrategy.updateRoadwaysAndLinearLocationsWithProjectLinks() When two track road with a new track 2 and partially new track 1 having new track in the middle Then formed roadways should be continuous. *** FAILED ***
Viite_13_218_spec:
- Test road_13_218 *** FAILED ***